src/bt_peer_connection.cpp

```diff
@@ -1230,6 +1139,231 @@ namespace {
+
+		auto const file_root = sha256_hash(ptr);
+		file_index_t file_index{-1};
+		for (file_index_t i : files.file_range())
+		{
+			if (files.root(i) == file_root)
+			{
+				file_index = i;
+				break;
+			}
+		}
+		ptr += sha256_hash::size();
+		int const base = detail::read_int32(ptr);
+		int const index = detail::read_int32(ptr);
+		int const count = detail::read_int32(ptr);
+		int const proof_layers = detail::read_int32(ptr);
+		hash_request hr(file_index, base, index, count, proof_layers);
+
+		if (!validate_hash_request(hr, t->torrent_file().files()))
+		{
+			write_hash_reject(hr);
+			return;
+		}
+
+#ifndef TORRENT_DISABLE_LOGGING
+		if (should_log(peer_log_alert::incoming_message))
+		{
+			peer_log(peer_log_alert::incoming_message, "HASH_REQUEST", "%d %d %d %d %d"
+				, static_cast<int>(hr.file), hr.base, hr.index, hr.count, hr.proof_layers);
+		}
+#endif
+
+		std::vector<sha256_hash> hashes = t->get_hashes(hr);
+
+		if (hashes.empty())
+		{
+			write_hash_reject(hr);
+			return;
+		}
+
+		write_hashes(hr, hashes);
+	}
+
+	void bt_peer_connection::on_hashes(int received)
+	{
+		INVARIANT_CHECK;
+
+		TORRENT_ASSERT(received >= 0);
+		received_bytes(0, received);
+
+		if (!peer_info_struct()->protocol_v2)
+		{
+			disconnect(errors::invalid_message, operation_t::bittorrent);
+			return;
+		}
+
+		std::shared_ptr<torrent> t = associated_torrent().lock();
+		TORRENT_ASSERT(t);
+
+		auto const& files = t->torrent_file().files();
+
+		span<char const> recv_buffer = m_recv_buffer.get();
+
+		int const header_size = 1 + 32 + 4 + 4 + 4 + 4;
+
+		if (recv_buffer.size() < header_size)
+		{
+			return;
+		}
+
+		const char* ptr = recv_buffer.begin() + 1;
+
+		auto const file_root = sha256_hash(ptr);
+		file_index_t file_index{ -1 };
+		for (file_index_t i : files.file_range())
+		{
+			if (files.root(i) == file_root)
+			{
+				file_index = i;
+				break;
+			}
+		}
+		ptr += sha256_hash::size();
+		int const base = detail::read_int32(ptr);
+		int const index = detail::read_int32(ptr);
+		int const count = detail::read_int32(ptr);
+		int const proof_layers = detail::read_int32(ptr);
+
+		hash_request const hr(file_index, base, index, count, proof_layers);
+
+		if (!validate_hash_request(hr, t->torrent_file().files()))
+		{
+			disconnect(errors::invalid_hashes, operation_t::bittorrent, peer_connection_interface::peer_error);
+			return;
+		}
+
+		// subtract one because the the base layer doesn't count
+		int const proof_hashes = std::max(0
+			, proof_layers - (merkle_num_layers(merkle_num_leafs(count)) - 1));
+
+		if (m_recv_buffer.packet_size() != header_size
+			+ (count + proof_hashes) * int(sha256_hash::size()))
+		{
+			disconnect(errors::invalid_hashes, operation_t::bittorrent, peer_connection_interface::peer_error);
+			return;
+		}
+
+		if (!m_recv_buffer.packet_finished()) return;
+
+		auto new_end = std::remove(m_hash_requests.begin(), m_hash_requests.end(), hr);
+		m_hash_requests.erase(new_end, m_hash_requests.end());
```

Comment:
I think there's a risk here, that `hr` is not in `m_hash_requests`, and that will prevent us from sending more than two requests. Maybe that's fine, because this is just a single peer, and eventually maybe we'll disconnect the peer. and we can always request from other peers, right?

In fact, looking into this, it seems like we (probably) could replace the `m_hash_requests` vector with just a counter of outstanding hash requests. The only difference (as far as I can tell) would be how we handle the error cases of peers sending responses that we didn't request. But it doesn't seem important that it's handled by not requesting any more from the peer.

Reply:
This code is modeled on what the ut_metadata plugin does. If we don't stop sending requests when a peer fails to respond to a request then there is the potential to get in a loop where we keep sending a request for the same hashes over-and-over again without making any progress towards completion.

In the event of a failed peer, `hash_picker::pick_hashes` will re-send the request to a different peer once all hashes have been requested at least once.